src/plugins/ui_feature/find_and_replace.ts

```diff
@@ -138,37 +154,119 @@ export class FindAndReplacePlugin extends UIPlugin {
     }
     this.currentSearchRegex = RegExp(searchValue, flags);
   }
-  /**
-   * Find matches using the current regex
-   */
+
+  // Find matches using the current regex
   private findMatches() {
-    const sheetId = this.getters.getActiveSheetId();
-    const cells = this.getters.getCells(sheetId);
     const matches: SearchMatch[] = [];
 
-    if (this.toSearch) {
-      for (const cell of Object.values(cells)) {
-        const { col, row } = this.getters.getCellPosition(cell.id);
-        if (
-          cell &&
-          this.currentSearchRegex &&
-          this.currentSearchRegex.test(this.getSearchableString({ sheetId, col, row }))
-        ) {
-          const match: SearchMatch = { col, row, selected: false };
-          matches.push(match);
+    if (!this.toSearch) {
+      return matches;
+    }
+
+    switch (this.searchOptions.searchScope) {
+      case "allSheets":
+        const sheets = this.getters.getSheetIds();
+        for (const [sheetIndex, sheet] of sheets.entries()) {
+          this.findMatchesInSheet(sheet, sheetIndex, matches);
         }
-      }
+        break;
+      case "thisSheet":
+        const sheetId = this.getters.getActiveSheetId();
+        const sheetIndex = this.getters.getSheetIds().indexOf(sheetId);
+        this.findMatchesInSheet(sheetId, sheetIndex, matches);
+        break;
+      case "specificRange":
+        const specificRange = this.searchOptions.specificRange;
+        if (specificRange) {
+          let rangeSheet, rangeCells;
+
+          if (specificRange.includes("!")) {
+            [rangeSheet, rangeCells] = specificRange.split("!");
+          } else {
+            rangeSheet = this.getters.getSheetName(this.getters.getActiveSheetId());
+            rangeCells = specificRange;
+          }
+
+          const [startCell, endCell] = rangeCells.split(":");
+
+          const finalEndCell = endCell || startCell;
+          const specificRangeWithEndCell = `${startCell}:${finalEndCell}`;
+
+          const sheets = this.getters.getSheetIds();
+          for (const [sheetIndex, sheet] of sheets.entries()) {
+            const sheetName = this.getters.getSheetName(sheet);
+            const cleanedRangeSheet = rangeSheet.replace(/'/g, "");
+            if (sheetName === cleanedRangeSheet) {
+              this.findMatchesInSheet(sheet, sheetIndex, matches, specificRangeWithEndCell);
+            }
+          }
+        }
+        break;
     }
-    return matches.sort(this.sortByRowThenColumn);
+
+    // Sort the matches by sheet index, row, and column
+    return matches.sort(this.sortBySheetIndexThenRowThenColumn);
   }
 
-  private sortByRowThenColumn(a: Position, b: Position) {
-    if (a.row === b.row) {
-      return a.col - b.col;
+  private findMatchesInSheet(
+    sheetId: string,
+    sheetIndex: number,
+    matches: SearchMatch[],
+    specificRange?: string
+  ) {
+    let top, left, bottom, right;
```

Comment:
Don't ever, unless you have no other choice, type variable as `any`. Your code can do random stuff, because `rangeData._zone.bottom` can be undefined, and you didn't notice it because you didn't type teh variable. The whole point of typescript is to avoid such errors.

Reply:
And if you're gonna use `top, left, right, bottom`, just directly use a `Zone`...